File: web/src/services/message-service.ts
/**
 * Message service
 * Handles message creation, SOS cooldown tracking, and message operations
 */

import type { Message, MessageCreateRequest, SOSType } from '../domain/message';
import { validateMessageCreateRequest } from '../domain/message';
import { generateId } from '../utils/id';
import { getDatabase } from './db';
import { getOrCreateDeviceId } from './device-storage';

// SOS cooldown duration (30 seconds)
const SOS_COOLDOWN_DURATION = 30 * 1000; // milliseconds

// Store last SOS timestamp per device (in localStorage for persistence)
const LAST_SOS_KEY_PREFIX = 'last_sos_';

/**
 * Gets the last SOS timestamp for a device
 * @param deviceId - Device ID
 * @returns Last SOS timestamp or null
 */
function getLastSOSTimestamp(deviceId: string): number | null {
  const key = `${LAST_SOS_KEY_PREFIX}${deviceId}`;
  const stored = localStorage.getItem(key);
  if (!stored) return null;
  return parseInt(stored, 10);
}

/**
 * Records that a device sent an SOS message
 * @param deviceId - Device ID
 */
function recordSOSMessage(deviceId: string): void {
  const key = `${LAST_SOS_KEY_PREFIX}${deviceId}`;
  localStorage.setItem(key, Date.now().toString());
}

/**
 * Checks if a device can send an SOS message (30 second cooldown)
 * @param deviceId - Device ID
 * @returns Error message if cooldown active, null if allowed
 */
export function checkSOSCooldown(deviceId: string): string | null {
  const lastSOS = getLastSOSTimestamp(deviceId);
  if (!lastSOS) {
    return null; // No previous SOS, allowed
  }

  const timeSinceLastSOS = Date.now() - lastSOS;
  if (timeSinceLastSOS < SOS_COOLDOWN_DURATION) {
    const remaining = Math.ceil((SOS_COOLDOWN_DURATION - timeSinceLastSOS) / 1000);
    return `SOS cooldown active: please wait ${remaining} more seconds`;
  }

  return null;
}

/**
 * Creates an SOS message
 * @param groupId - Group ID
 * @param sosType - SOS type
 * @param content - Optional message content
 * @returns Created message
 */
async function persistMessage(payload: Message): Promise<Message> {
  const db = await getDatabase();
  await db.messages.insert(payload);
  return payload;
}

export async function createMessage(request: MessageCreateRequest): Promise<Message> {
  const deviceId = getOrCreateDeviceId();
  const validationError = validateMessageCreateRequest(request);
  if (validationError) {
    throw new Error(validationError);
  }

  const message: Message = {
    id: generateId(),
    group_id: request.group_id,
    device_id: deviceId,
    content: request.content,
    message_type: request.message_type,
    sos_type: request.sos_type,
    tags: request.tags,
    pinned: false,
    created_at: new Date().toISOString(),
    device_sequence: request.device_sequence,
    sync_status: 'pending',
  };

  return persistMessage(message);
}

export async function createSOSMessage(
  groupId: string,
  sosType: SOSType,
  content?: string
): Promise<Message> {
  const deviceId = getOrCreateDeviceId();
  const cooldownError = checkSOSCooldown(deviceId);
  if (cooldownError) {
    throw new Error(cooldownError);
  }

  const message = await createMessage({
    group_id: groupId,
    content: content || getSOSDefaultContent(sosType),
    message_type: 'sos',
    sos_type: sosType,
    tags: ['urgent'],
  });

  recordSOSMessage(deviceId);
  return message;
}

/**
 * Gets default content for SOS message based on type
 * @param sosType - SOS type
 * @returns Default message content
 */
function getSOSDefaultContent(sosType: SOSType): string {
  const contentMap: Record<SOSType, string> = {
    medical: '🚨 Medical Emergency - Need immediate help!',
    flood: '🚨 Flood Emergency - Need evacuation assistance!',
    fire: '🚨 Fire Emergency - Need immediate help!',
    missing_person: '🚨 Missing Person - Need help locating!',
  };
  return contentMap[sosType] || '🚨 Emergency - Need help!';
}

/**
 * Creates a regular text message
 * @param groupId - Group ID
 * @param content - Message content
 * @returns Created message
 */
export async function createTextMessage(groupId: string, content: string): Promise<Message> {
  return createMessage({
    group_id: groupId,
    content,
    message_type: 'text',
  });
}

/**
 * Gets unread message count for a group
 * Unread = messages created after the last time user viewed the group
 * For now, we'll use a simplified approach: count messages user hasn't sent
 * @param groupId - Group ID
 * @param lastReadAt - Optional timestamp of last read (ISO string)
 * @returns Unread message count
 */
export async function getUnreadCount(
  groupId: string,
  lastReadAt?: string
): Promise<number> {
  const db = await getDatabase();
  const deviceId = getOrCreateDeviceId();

  // Build query selector
  const selector: {
    group_id: string;
    device_id: { $ne: string };
    created_at?: { $gt: string };
  } = {
    group_id: groupId,
    device_id: { $ne: deviceId }, // Exclude own messages
  };

  // If lastReadAt is provided, only count messages after that time
  if (lastReadAt) {
    selector.created_at = { $gt: lastReadAt };
  }

  const unreadMessages = await db.messages.find({ selector }).exec();
  return unreadMessages.length;
}

File: api/internal/service/message_service.go
package service

import (
	"context"
	"fmt"
	"time"

	"nearby-msg/api/internal/domain"
	"nearby-msg/api/internal/infrastructure/database"
	"nearby-msg/api/internal/utils"
)

// MessageService handles message business logic
type MessageService struct {
	// Store last SOS timestamp per device (in-memory for now)
	// In production, this should be in Redis or database
	lastSOSTimestamps map[string]time.Time
	messageRepo       *database.MessageRepository
}

// NewMessageService creates a new message service
func NewMessageService(messageRepo *database.MessageRepository) *MessageService {
	return &MessageService{
		lastSOSTimestamps: make(map[string]time.Time),
		messageRepo:       messageRepo,
	}
}

// SOSCooldownDuration is the minimum time between SOS messages (30 seconds)
const SOSCooldownDuration = 30 * time.Second

// CheckSOSCooldown checks if a device can send an SOS message (30 second cooldown)
func (s *MessageService) CheckSOSCooldown(ctx context.Context, deviceID string) error {
	lastSOS, exists := s.lastSOSTimestamps[deviceID]
	if !exists {
		return nil // No previous SOS, allowed
	}

	timeSinceLastSOS := time.Since(lastSOS)
	if timeSinceLastSOS < SOSCooldownDuration {
		remaining := SOSCooldownDuration - timeSinceLastSOS
		return fmt.Errorf("SOS cooldown active: please wait %d more seconds", int(remaining.Seconds())+1)
	}

	return nil
}

// RecordSOSMessage records that a device sent an SOS message
func (s *MessageService) RecordSOSMessage(ctx context.Context, deviceID string) {
	s.lastSOSTimestamps[deviceID] = time.Now()
}

// CreateMessageRequest represents a message creation request
type CreateMessageRequest struct {
	GroupID        string             `json:"group_id"`
	DeviceID       string             `json:"device_id"`
	Content        string             `json:"content"`
	MessageType    domain.MessageType `json:"message_type"`
	SOSType        *domain.SOSType    `json:"sos_type,omitempty"`
	Tags           []string           `json:"tags,omitempty"`
	DeviceSequence *int               `json:"device_sequence,omitempty"`
}

// CreateMessage creates a new message with validation
func (s *MessageService) CreateMessage(ctx context.Context, req CreateMessageRequest) (*domain.Message, error) {
	// Check SOS cooldown if this is an SOS message
	if req.MessageType == domain.MessageTypeSOS {
		if err := s.CheckSOSCooldown(ctx, req.DeviceID); err != nil {
			return nil, err
		}
	}

	// Create message
	messageID, err := utils.GenerateID()
	if err != nil {
		return nil, fmt.Errorf("failed to generate message ID: %w", err)
	}

	message := &domain.Message{
		ID:             messageID,
		GroupID:        req.GroupID,
		DeviceID:       req.DeviceID,
		Content:        req.Content,
		MessageType:    req.MessageType,
		SOSType:        req.SOSType,
		Tags:           req.Tags,
		Pinned:         false,
		CreatedAt:      time.Now(),
		DeviceSequence: req.DeviceSequence,
	}

	// Validate message
	if err := message.Validate(); err != nil {
		return nil, err
	}

	// Record SOS if applicable
	if req.MessageType == domain.MessageTypeSOS {
		s.RecordSOSMessage(ctx, req.DeviceID)
	}

	return message, nil
}

// ValidateMessage validates a message
func (s *MessageService) ValidateMessage(message *domain.Message) error {
	return message.Validate()
}

// EnforceRetention ensures groups do not exceed the maximum number of messages.
func (s *MessageService) EnforceRetention(ctx context.Context, groupID string, maxMessages int) error {
	if s.messageRepo == nil {
		return nil
	}
	return s.messageRepo.TrimOldMessages(ctx, groupID, maxMessages)
}
